service/internal/db/db.go

```diff
@@ -186,3 +205,362 @@ func (d SQLClient) ListTokenEntries(ctx context.Context, project string) ([]Toke
+	metadataSK   = "METADATA"
+	tokenSKFmt   = "TOKEN#%s"
+)
+
+var (
+	ErrProjectNotFound = fmt.Errorf("project not found")
+	ErrTokenNotFound   = fmt.Errorf("token not found")
+)
+
+func NewDynamoDBClient(tableName string, endpointURL string, assumeRoleARN string) (*DynamoDBClient, error) {
+	var cfg aws.Config
+	var err error
+
+	if assumeRoleARN != "" {
+		cfg, err = config.LoadDefaultConfig(context.Background(),
+			config.WithAssumeRoleCredentialOptions(func(options *stscreds.AssumeRoleOptions) {
+				options.RoleARN = assumeRoleARN
+				options.RoleSessionName = "cello-session"
+			}),
+		)
+	} else {
+		cfg, err = config.LoadDefaultConfig(context.Background())
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to load AWS config: %w", err)
+	}
+
+	svc := dynamodb.NewFromConfig(cfg, func(o *dynamodb.Options) {
+		if endpointURL != "" {
+			o.BaseEndpoint = aws.String(endpointURL)
+		}
+	})
+
+	return &DynamoDBClient{
+		svc:       svc,
+		tableName: tableName,
+	}, nil
+}
+
+func (d *DynamoDBClient) Health(ctx context.Context) error {
+	// No-op implementation to avoid unnecessary API calls
+	return nil
+}
+
+func (d *DynamoDBClient) CreateProjectEntry(ctx context.Context, pe ProjectEntry) error {
+	item := map[string]ddbtypes.AttributeValue{
+		primaryKey:   &ddbtypes.AttributeValueMemberS{Value: fmt.Sprintf(projectPKFmt, pe.ProjectID)},
+		sortKey:      &ddbtypes.AttributeValueMemberS{Value: metadataSK},
+		"repository": &ddbtypes.AttributeValueMemberS{Value: pe.Repository},
+	}
+
+	_, err := d.svc.PutItem(ctx, &dynamodb.PutItemInput{
+		TableName:           aws.String(d.tableName),
+		Item:                item,
+		ConditionExpression: aws.String("attribute_not_exists(pk)"),
+	})
+	if err != nil {
+		var ccf *ddbtypes.ConditionalCheckFailedException
+		if errors.As(err, &ccf) {
+			return fmt.Errorf("project %s already exists", pe.ProjectID)
+		}
+		return fmt.Errorf("failed to create project: %w", err)
+	}
+	return nil
+}
+
+func (d *DynamoDBClient) ReadProjectEntry(ctx context.Context, project string) (ProjectEntry, error) {
+	result, err := d.svc.GetItem(ctx, &dynamodb.GetItemInput{
+		TableName: aws.String(d.tableName),
+		Key: map[string]ddbtypes.AttributeValue{
+			primaryKey: &ddbtypes.AttributeValueMemberS{Value: fmt.Sprintf(projectPKFmt, project)},
+			sortKey:    &ddbtypes.AttributeValueMemberS{Value: metadataSK},
+		},
+	})
+	if err != nil {
+		return ProjectEntry{}, err
+	}
+
+	if result.Item == nil {
+		return ProjectEntry{}, ErrProjectNotFound
+	}
+
+	repo, ok := result.Item["repository"].(*ddbtypes.AttributeValueMemberS)
+	if !ok {
+		return ProjectEntry{}, fmt.Errorf("invalid repository attribute")
+	}
+
+	return ProjectEntry{
+		ProjectID:  project,
+		Repository: repo.Value,
+	}, nil
+}
+
+func (d *DynamoDBClient) DeleteProjectEntry(ctx context.Context, project string) error {
+	projectPK := fmt.Sprintf(projectPKFmt, project)
+
+	// First, query all items associated with this project (metadata, tokens, etc.)
```

Comment:
have a look at how we handle this. for projects with many targets, can't do this in one go